python/ray/experimental/queue.py

```diff
@@ -45,70 +43,49 @@ def full(self):
     def put(self, item, block=True, timeout=None):
         """Adds an item to the queue.
 
-        Uses polling if block=True, so there is no guarantee of order if
-        multiple producers put to the same full queue.
+        There is no guarantee of order if multiple producers put to the same
+        full queue.
 
         Raises:
             Full if the queue is full and blocking is False.
+            Full if the queue is full, blocking is True, and it timed out.
+            ValueError if timeout is negative.
         """
-        if self.maxsize <= 0:
-            self.actor.put.remote(item)
-        elif not block:
-            if not ray.get(self.actor.put.remote(item)):
+        if not block:
+            try:
+                ray.get(self.actor.put_nowait.remote(item))
+            except asyncio.QueueFull:
                 raise Full
-        elif timeout is None:
-            # Polling
-            # Use a not_full condition variable or promise?
-            while not ray.get(self.actor.put.remote(item)):
-                # Consider adding time.sleep here
-                pass
-        elif timeout < 0:
-            raise ValueError("'timeout' must be a non-negative number")
         else:
-            endtime = time.time() + timeout
-            # Polling
-            # Use a condition variable or switch to promise?
-            success = False
-            while not success and time.time() < endtime:
-                success = ray.get(self.actor.put.remote(item))
-            if not success:
-                raise Full
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                ray.get(self.actor.put.remote(item, timeout))
```

Comment:
I think there's a problem here - if the timeout is reached, we stop the `ray.get` call locally, but the remote task to `self.actor.put.remote` will still continue. In order to provide the behavior that it either succeeds (is in the queue) or times out, you'll need to call a `ray.cancel` here and check to see if it was successfully cancelled.

Reply:
I think what you are saying would be true if the line were `ray.get(_, timeout)` .  This was my older approach, which failed the existing asynchronous usage test (for the reason you describe).  

In the present approach, the timeout parameter is inside the `actor.put()` call.  The `ray.get()` doesn't get stopped.  Inside `actor.put(_, timeout)` there is an `await asyncio.wait_for(_, timeout)`, so when the time out is reached, it catches the asyncio timeout error and raises `Empty` as desired.